SCAutolib/models/card.py

```diff
@@ -0,0 +1,229 @@
+(physical) smart card in standard reader, cards in the removinator.
+"""
+import re
+import time
+from pathlib import Path
+from traceback import format_exc
+
+from SCAutolib import run, logger, TEMPLATES_DIR
+from SCAutolib.models.file import SoftHSM2Conf
+from SCAutolib.models.user import User
+
+
+class Card:
+    """
+    Interface for child classes. All child classes will rewrite common methods
+    based on the type of the card.
+    """
+    uri: str = None
+    user: User = None
+    _pattern: str = None
+
+    def _set_uri(self):
+        """
+        Sets URI for given smart card. Uri is matched from ``p11tool`` command
+        with regular expression. If URI is not matched, assertion exception
+        would be raised
+
+        :raise: AssertionError
+        """
+        cmd = ["p11tool", "--list-token-urls"]
+        out = run(cmd).stdout
+        urls = re.findall(self._pattern, out)
+        assert len(urls) == 1, "More URLs are matched"
+        self.uri = urls[0]
+
+    def insert(self):
+        """
+        Insert the card.
+        """
+        ...
+
+    def remove(self):
+        """
+        Remove the card.
+        """
+        ...
+
+    def enroll(self):
+        """
+        Enroll the card (upload a certificate and a key on it)
+        """
+        ...
+
+
+class VirtualCard(Card):
+    """
+    This class provides method for manipulating with virtual smart card. Virtual
+    smart card by itself is represented by the systemd service in the system.
+    The card relates to some user, so providing the user is essential for
+    correct functioning of methods for the virtual smart card.
+
+    Card root directory has to be created before calling any method
+    """
+
+    _service_name: str = None
+    _service_location: Path = None
+    _softhsm2_conf: Path = None
+    _nssdb: Path = None
+    _template: Path = Path(TEMPLATES_DIR, "virt_cacard.service")
+    _pattern = r"(pkcs11:model=PKCS%2315%20emulated;" \
+               r"manufacturer=Common%20Access%20Card;serial=.*)"
+
+    def __init__(self, cert: Path, key: Path, user: User, insert: bool = False,
+                 softhsm2_conf: Path = None):
+        """
+        Initialise virtual smart card. Constructor of the base class is also
+        used.
+
+        :param cert: Path to the certificate for this card
+        :type cert: pathlib.Path
+        :param key: Path to private key for this card
+        :type key: pathlib.Path
+        :param user: User of this card
+        :type user: User
+        :param insert: If True, the card would be inserted on entering the
+            context manger. Default False.
+        :type insert: bool
+        :param softhsm2_conf: path to SoftHSM2 configuration file
+        :type softhsm2_conf: pathlib.Path
+        """
+
+        self.user = user
+        assert self.user.card_dir.exists(), "Card root directory doesn't exists"
+
+        self._private_key = key
+        self._cert = cert
```

Comment:
The user needs to be created before we can create a virtual card, so instead of providing them here we can just use the cert and key directly from the user, no?